### PR TITLE
Add linear_bandit.py supporting UCB and epsilon greedy exploration

### PR DESCRIPTION
**Pull request by posgnu**
### Issue 

https://github.com/probml/pyprobml/issues/699

## Checklist:

- [x] Performed a self-review of the code
- [x] Tested on Google Colab.

## Experiment results

https://github.com/probml/pyprobml/issues/699#issuecomment-1086777028

### CODE REVIEW

**Comment by posgnu** [comment]
@murphyk @patel-zeel 

Could you review my implementation?

 I am not sure this is the right place to put the code so I would appreciate it if you could inform me of it.

**Comment by murphyk** [comment]
LGTM.

File: scripts/linear_bandit.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            return self._epsilon
+
+    def init_bel(self, contexts, actions, rewards):
+        mu = jnp.zeros((self.num_arms, self.num_features))
+        Sigma = (
+            1.0
+            / self.lmbda
+            * jnp.eye(self.num_features)
+            * jnp.ones((self.num_arms, 1, 1))
+        )
```
Comment (gerdm) [inline scripts/linear_bandit.py:60]: We could rewrite this as follows:

```
Sigma = jnp.eye(self.num_features) * jnp.ones((self.num_arms, 1, 1) / self.lmbda
```

Or even better:
```
Sigma = jnp.ones((self.num_arms, self.num_features, self.num_features)) / self.lmbda
```

File: scripts/linear_bandit.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    def __init__(
+        self,
+        num_features,
+        num_arms,
+        exploration_policy,
+        eta=6.0,
+        lmbda=0.25,
+        alpha=1.0,
+        epsilon=0,
+    ):
```
Comment (gerdm) [inline scripts/linear_bandit.py:30]: No need for the extra breaks. We could refactor this as:

```
def __init__(self,  num_features, num_arms, exploration_policy,
             eta=6.0, lmbda=0.25, alpha=1.0, epsilon=0):
```

File: scripts/linear_bandit.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        a_update = a_k + 1 / 2
+        b_update = (
+            b_k
+            + (
+                reward ** 2
+                + mu_k.T @ Lambda_k @ mu_k
+                - mu_update.T @ Lambda_update @ mu_update
+            )
+            / 2
+        )
```
Comment (gerdm) [inline scripts/linear_bandit.py:98]: Don't break lines too often.

**Review by gerdm — COMMENTED** [review]
Code looks good, I would only suggest a couple of changes to the style of the code.

File: scripts/linear_bandit.py (hunk 3 above)
Comment (gerdm) [reply scripts/linear_bandit.py:98]: See comment above.

File: scripts/linear_bandit.py (hunk 1 above)
Comment (posgnu) [reply scripts/linear_bandit.py:60]: I like the last one. Thanks

File: scripts/linear_bandit.py (hunk 2 above)
Comment (posgnu) [reply scripts/linear_bandit.py:30]: Okay, it seems auto-formatting using Black sometimes seems messier.

File: scripts/linear_bandit.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        action = predicted_reward.argmax()
+        return action
+
+    def ucb_policy(self, key, bel, context):
+        mu, Sigma, a, b = bel
+        covariance_matrix = Sigma * Sigma
+        predicted_reward = jnp.einsum("m,km->k", context, mu)
+        predicted_variance = jnp.einsum(
+            "kl,l-> k", jnp.einsum("m,kml->kl", context, covariance_matrix), context
+        )
```
[the remark this answers is not in the document]
Comment (posgnu) [reply scripts/linear_bandit.py:133]: I am not sure it is worth making it hard to read and add comments instead.

[the remark this answers is not in the document]
Comment (gerdm) [reply scripts/linear_bandit.py:133]: The outer einsum is a matrix vector multiplication, the inner-most einsum is another vector matrix multiplication over `K` covariance matrices. We obtain: 
```
jnp.einsum("n,knm,m->k", context, covs, context)
```

[the remark this answers is not in the document]
Comment (posgnu) [reply scripts/linear_bandit.py:133]: That makes sense.

File: scripts/linear_bandit.py (hunk 1 above)
Comment (gerdm) [reply scripts/linear_bandit.py:60]: You're totally right! Thanks for pointing that out @posgnu.

**Comment by posgnu** [comment]
@gerdm I updated the PR.

**Pull request merged**